cpp/include/cuspatial/detail/find/find_and_combine_segment.cuh

```diff
@@ -57,20 +62,66 @@ void __global__ simple_find_and_combine_segments_kernel(OffsetRange offsets,
   }
 }
 
+/**
+ * @brief Comparator for sorting the segment range.
+ *
+ * This comparator makes sure that the segment range are sorted by the following keys:
+ * 1. Segments with the same space id are grouped together.
+ * 2. Segments within the same space are grouped by their slope.
+ * 3. Within each slope group, segments are sorted by their lower left point.
+ */
+template <typename index_t, typename T>
+struct segment_comparator {
+  bool __device__ operator()(thrust::tuple<index_t, segment<T>> const& lhs,
+                             thrust::tuple<index_t, segment<T>> const& rhs) const
+  {
+    auto lhs_index   = thrust::get<0>(lhs);
+    auto rhs_index   = thrust::get<0>(rhs);
+    auto lhs_segment = thrust::get<1>(lhs);
+    auto rhs_segment = thrust::get<1>(rhs);
+
+    // Compare space id
+    if (lhs_index == rhs_index) {
+      // Compare slope
+      if (lhs_segment.collinear(rhs_segment)) {
+        // Sort by the lower left point of the segment.
+        return lhs_segment.lower_left() < rhs_segment.lower_left();
+      }
+      return lhs_segment.slope() < rhs_segment.slope();
+    }
+    return lhs_index < rhs_index;
+  }
+};
+
 /**
  * @internal
  * @brief For each pair of mergeable segment, overwrites the first segment with merged result,
  * sets the flag for the second segment as 1.
+ *
+ * @note This function will alter the input segment range by rearranging the order of the segments
+ * within each space so that merging kernel can take place.
  */
 template <typename OffsetRange, typename SegmentRange, typename OutputIt>
 void find_and_combine_segment(OffsetRange offsets,
                               SegmentRange segments,
                               OutputIt merged_flag,
                               rmm::cuda_stream_view stream)
 {
+  using index_t   = typename OffsetRange::value_type;
+  using T         = typename SegmentRange::value_type::value_type;
   auto num_spaces = offsets.size() - 1;
   if (num_spaces == 0) return;
 
+  // Construct a key iterator using the offsets of the segment and the segment itself.
+  auto space_id_iter         = make_geometry_id_iterator<index_t>(offsets.begin(), offsets.end());
+  auto space_id_segment_iter = thrust::make_zip_iterator(space_id_iter, segments.begin());
+
+  thrust::sort_by_key(rmm::exec_policy(stream),
```

Comment:
What is the impact of the new sorting on performance of linestring intersection?

Reply:
Interestingly, because the sorting will make `find_and_merge_segment` kernel avoid computing many segments in the inner loop scan, in some cases the performance actually improved. In worst case scenario the performance decreases by 3.4%.

Ref Time: branch-23.08, Cmp Time: this branch.
```
|  FP type  |  num_pairs  |  num_linestring_per_multilinestring  |  num_segments_per_linestring  |   Ref Time |   Ref Noise |   Cmp Time |   Cmp Noise |          Diff |   %Diff |  Status  |
|-----------|-------------|--------------------------------------|-------------------------------|------------|-------------|------------|-------------|---------------|---------|----------|
|    F32    |    1000     |                  1                   |              100              |  11.537 ms |       2.38% |  11.930 ms |       2.33% |    393.774 us |   3.41% |   FAIL   |
|    F32    |    10000    |                  1                   |              100              |  12.108 ms |       2.83% |  12.042 ms |       5.02% |    -66.531 us |  -0.55% |   PASS   |
|    F32    |   100000    |                  1                   |              100              |  28.847 ms |       0.22% |  27.647 ms |       0.28% |  -1200.119 us |  -4.16% |   FAIL   |
|    F32    |   1000000   |                  1                   |              100              | 275.615 ms |       0.05% | 248.693 ms |       0.07% | -26922.072 us |  -9.77% |   FAIL   |
|    F64    |    1000     |                  1                   |              100              |  15.375 ms |       1.36% |  15.559 ms |       1.34% |    183.851 us |   1.20% |   PASS   |
|    F64    |    10000    |                  1                   |              100              |  17.503 ms |       2.20% |  17.775 ms |       1.83% |    272.187 us |   1.56% |   PASS   |
|    F64    |   100000    |                  1                   |              100              |  55.969 ms |       2.20% |  56.857 ms |       1.82% |    887.096 us |   1.58% |   PASS   |
|    F64    |   1000000   |                  1                   |              100              | 431.700 ms |       0.30% | 430.470 ms |       0.29% |  -1230.114 us |  -0.28% |   PASS   |
```